### ISSUE TITLE
One question about code in derivatives of the AdCo Loss

### ISSUE THREAD

**Issue by FiresWorker** [maple-research-lab/AdCo#8]
I have a question about the code in updating the memory bank:
g = torch.einsum('cn,nk->ck',[q.T,p_qd])/logits.shape[0] - torch.mul(torch.mean(torch.mul(p_qd,l_neg),dim=0),d_norm).

why do we compute the "torch.mul(torch.mean(torch.mul(p_qd,l_neg),dim=0),d_norm)" and what is its meaning.

Thank you very much for answering my question.

**Comment by WoodwindHu** [comment]
In the paper, the gradient of n_k to L is:
![image](https://user-images.githubusercontent.com/28439907/121285722-6e794380-c911-11eb-9083-0a44f2300c05.png)
But we stored unnormalized negative adversaries in the memory bank. And when we calculate the gradient of n_ku (unnormalized) to L, there is another term.  
![image](https://user-images.githubusercontent.com/28439907/121287576-653da600-c914-11eb-80f4-55ffab7d5bd8.png)
![image](https://user-images.githubusercontent.com/28439907/121287388-06782c80-c914-11eb-94f2-d4b4aee91261.png)
The second term in the gradient is the meaning the "torch.mul(torch.mean(torch.mul(p_qd,l_neg),dim=0),d_norm)".
